### PR TITLE
Improve UserObjectInterface errors

### LINKED ISSUE #17446: Add API to get fluid properties
## Reason
There are a number of different base classes in the fluid properties module that the fluid property user objects derive from - such as `SinglePhaseFluidProperties` and `TwoPhaseFluidProperties`. Various MOOSE apps may place specific requirements on which of these fluid property options are valid - for instance, Pronghorn restricts fluid properties to those deriving from `SinglePhaseFluidProperties`. However, to fully check that a user does not try to use a different fluid property base in Pronghorn, we need to downcast a `FluidProperties *` userobject in each and every Pronghorn class that uses fluid properties to make sure that the correct type was provided:

```
const FluidProperties * what_user_provided = getUserObject<GeneralUserObject>("fluid");
const SinglePhaseFluidProperties fluid = dynamic_cast<const SinglePhaseFluidProperties *>(what_user_provided);
if (!fluid)
  mooseError("Pronghorn needs SinglePhaseFluidProperties");
```

This is tedious and open to errors if you forget to perform this check in new classes. Further, the downcasting makes us use pointers, whereas MOOSE would like all MOOSE-based apps to use references whenever possible.

## Design
Add a free function like `getFluidProperties<T>("name")` that automatically checks whether the correct fluid property type has been provided by the user, and error if not.

## Impact
Better error checking and protecting Pronghorn users from accidentally using fluid properties that are incompatible with Pronghorn. Benefits to all other applications using fluid properties as well.

### PR DESCRIPTION
**Pull request by aprilnovak**
Small modification to `UserObjectInterface` to throw an error if the type of the user object being grabbed does not match `T`, as well as a test demonstrating the desired behavior for the fluid properties module. 

After discussion with @andrsd, it may be more useful for apps and other users to have this patch directly in `UserObjectInterface`(as opposed to a new `FluidPropertyInterface`) because users no longer would need to modify their source files to take advantage of this check. Plus, this error check is pretty general and would likely be useful to all MOOSE objects with user objects.

I wanted to print the type of `T` to get an even better error message, with

```
T * dummy;
typeid(dummy).name()
```

but apparently that's implementation-dependent. So if there's some way to figure that out in MOOSE, I think that'd be great to add to the error message.

Closes #17446 and refs #17512

### CODE REVIEW

File: framework/include/userobject/UserObjectInterface.h [hunk 1]
```diff
@@ -62,6 +62,9 @@ class UserObjectInterface
   const UserObject & getUserObjectBaseByName(const std::string & name) const;
 
 private:
+  /// Moose object using the user object
+  const MooseObject * _moose_object;
```
Comment (loganharbour) [inline framework/include/userobject/UserObjectInterface.h:66]:
```suggestion
  const MooseObject * const _uoi_moose_object;
```

File: framework/include/userobject/UserObjectInterface.h [hunk 2]
```diff
@@ -80,7 +83,18 @@ const T &
 UserObjectInterface::getUserObject(const std::string & name) const
 {
   unsigned int tid = needThreadedCopy(getUserObjectBase(name)) ? _uoi_tid : 0;
-  return _uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name), tid);
+  const auto * uo = &_uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name), tid);
```
Comment (loganharbour) [inline framework/include/userobject/UserObjectInterface.h:86]:
```suggestion
  auto uo = &_uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name), tid);
```

File: framework/include/userobject/UserObjectInterface.h [hunk 3]
```diff
@@ -80,7 +83,18 @@ const T &
...
+
+  if (!dynamic_cast<const T *>(uo))
+    _moose_object->paramError(name,
```
Comment (loganharbour) [inline framework/include/userobject/UserObjectInterface.h:89]:
```suggestion
    _uoi_moose_object->paramError(name,
```

File: framework/include/userobject/UserObjectInterface.h [hunk 4]
```diff
@@ -80,7 +83,18 @@ const T &
...
+                              "UserObject '" + _uoi_params.get<UserObjectName>(name) +
+                                  "' does not derive from " +
+                                  "the base class that is expected by " + _moose_object->type() +
+                                  ".\nPlease consult "
+                                  "the documentation for " +
+                                  _moose_object->type() + ".");
```
Comment (loganharbour) [inline framework/include/userobject/UserObjectInterface.h:95]:
```suggestion
                              "The supplied UserObject of type ", uo->type(), " does not derive from the base class expected (", typeid(T).name(), ")."
```

How about something like this?

File: framework/src/userobject/UserObjectInterface.C [hunk 5]
```diff
@@ -14,7 +14,8 @@
 #include "InputParameters.h"
 
 UserObjectInterface::UserObjectInterface(const MooseObject * moose_object)
-  : _uoi_params(moose_object->parameters()),
+  : _moose_object(moose_object),
```
Comment (loganharbour) [inline framework/src/userobject/UserObjectInterface.C:17]:
```suggestion
  : _uoi_moose_object(moose_object),
```

**Review by loganharbour — DISMISSED** [review]

File: framework/include/userobject/UserObjectInterface.h (hunk 4 above)
Comment (aprilnovak) [reply framework/include/userobject/UserObjectInterface.h:95]: Unfortunately `typeid` is [implementation-dependent](https://en.cppreference.com/w/cpp/types/type_info/name), so what that prints can vary from useful to not-useful. For me, it prints:

```
P26SinglePhaseFluidProperties
```

for the test I added, where the 26 I've gleaned is the number of characters in the type name, and idk what P is.

Comment (loganharbour) [reply framework/include/userobject/UserObjectInterface.h:95]: Ah yeah, I remember that. I typed this up without looking at any of the doco. I swore we had an implementation somewhere...

**Comment by loganharbour** [comment]
Okay, this has led me down a path of frustration with this interface. It's one of those Sunday-morning-clean-up-MOOSE mornings. Will push a rewrite by this afternoon.

**Comment by loganharbour** [comment]
@aprilnovak Let me know what you think:

```
*** ERROR ***
/Users/harblh/projects/moose/test/tests/interfaces/userobjectinterface/uoi_errors.i:14: (UserObjects/error_test/uo):
     The provided object "null_uo" of type NullUserObject is not derived from the required type.

     The object must derive from ThreadedGeneralUserObject.
```

**Comment by aprilnovak** [comment]
I like it! That's perfect - thanks for the input :)

**Comment by loganharbour** [comment]
Unfortunately... this has led me down a swath of other error fixing and cleaning. I'll take a break for now and push what I have so we can get this in.

The errors with the previous implementation were that the lower level interfaces (FEProblemBase::getUserObject<T>) do the casting ahead of time when getting the object. Therefore, if they don't match, they just return nothing

File: framework/src/userobject/UserObjectInterface.C [hunk 6]
```diff
@@ -7,34 +7,86 @@
...
+  return _uoi_feproblem.getUserObjectBase(object_name, tid);
 }
 
 bool
 UserObjectInterface::needThreadedCopy(const UserObject & uo) const
 {
-  return (dynamic_cast<const DiscreteElementUserObject *>(&uo) != NULL) ||
-         (dynamic_cast<const ThreadedGeneralUserObject *>(&uo) != NULL);
+  return dynamic_cast<const DiscreteElementUserObject *>(&uo) ||
+         dynamic_cast<const ThreadedGeneralUserObject *>(&uo);
```
Comment (rwcarlsen) [inline framework/src/userobject/UserObjectInterface.C:79]: I know the implementation of this function isn't being changed here - but it seems to me like this function should return true if `uo` has threaded copies - and there are several user object types for which this is the case that aren't checked here - i.e. ElementUserObject subclasses, etc.  Does this need to be fixed?

**Review by rwcarlsen — CHANGES_REQUESTED** [review]

File: framework/include/auxkernels/AuxKernel.h [hunk 7]
```diff
@@ -119,37 +119,40 @@ class AuxKernelTempl : public MooseObject,
   const MaterialProperty<T> & getMaterialPropertyOlder(const std::string & name);
 
   template <typename T>
-  const T & getUserObject(const UserObjectName & name);
+  const T & getUserObject(const std::string & param_name) const;
```
Comment (andrsd) [inline framework/include/auxkernels/AuxKernel.h:122]: :+1: for this and the others

File: framework/src/userobject/UserObjectInterface.C (hunk 6 above)
Comment (loganharbour) [reply framework/src/userobject/UserObjectInterface.C:79]: I actually thought about putting something like
```
virtual threaded() const { return false; }
```
in UserObject, and overriding it where needed and just using that instead. Thoughts?

File: framework/include/auxkernels/AuxKernel.h (hunk 7 above)
Comment (loganharbour) [reply framework/include/auxkernels/AuxKernel.h:122]: There's still lots to be done for the PP and VPP interfaces. Let me make an issue...

File: framework/src/userobject/UserObjectInterface.C (hunk 6 above)
Comment (rwcarlsen) [reply framework/src/userobject/UserObjectInterface.C:98]: Yeah - that would be better than the casting nonsense.

**Review by rwcarlsen — CHANGES_REQUESTED** [review]

**Review by rwcarlsen — APPROVED** [review]

**Comment by pbalest** [comment]
@sallustius  and I, we are working on mixture properties and we had to write functions like p_from_blabla_and_x that inherits from fluid properties and not singlephase fluid properties, this "NICE" feature is blocking our development. Any quick fix??? We are up for a delivarable soon any quick suggestion is very much appreciated.

**Comment by loganharbour** [comment]
@pbalest I'll ping you in a message

**Pull request merged**